src/System.Management.Automation/engine/InitialSessionState.cs

```diff
@@ -4822,6 +4822,184 @@ .ForwardHelpCategory Cmdlet
 ";
         }
 
+#if UNIX
+        internal static string GetSudoFunctionText()
+        {
+            return @"
+Set-StrictMode -Version Latest
+
+$script:AmbiguousOperators = @('>', '<', '|', '&')
+$script:IgnoredTokens = @(
+    [System.Management.Automation.PSTokenType]::CommandParameter,
+    [System.Management.Automation.PSTokenType]::CommandArgument,
+    [System.Management.Automation.PSTokenType]::NewLine,
+    [System.Management.Automation.PSTokenType]::Comment,
+    [System.Management.Automation.PSTokenType]::StatementSeparator,
+    [System.Management.Automation.PSTokenType]::LineContinuation
+    )
+
+function Get-SudoArgs
+{
+    param
+    (
+        [System.Collections.Generic.List[string]] $AllArgs
+    )
+    $sb = [System.Text.StringBuilder]::new()
+
+    [string] $arg = $AllArgs[0]
+    while ($arg.StartsWith('-'))
+    {
+        $AllArgs.RemoveAt(0)
+        $null = $sb.Append(' ')
+        $null = $sb.Append($arg)
+
+        if ($AllArgs.Count -gt 0)
+        {
+            $arg = $AllArgs[0]
+        }
+        else
+        {
+            break
+        }
+    }
+    return $sb.ToString()
+}
+
+function Get-ScriptBlock
+{
+    param
+    (
+        [Parameter(Mandatory)]
+        [System.Collections.Generic.List[string]] $AllArgs,
+
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string] $SudoCmd
+    )
+    [bool] $usePowerShell = $false
+
+    [string] $sudoArgs = Get-SudoArgs -AllArgs $AllArgs
+
+    <# tokenize the remaining arguments to identify the type of expression; native or powershell
```

Comment:
I understand but that presents a different question/problem and I'm not conversant with the parser to know how it can resolve it.
If I tokenize the string $x + $x, $x get's resolved to echo.  If I enclose the same expression in a scriptblock, $x is parsed as a variable reference.  My immediate thought is to resolve these variable references and use the resulting expression.  However, this presents a different problem, what if the expression is an assignment, such as the following:
$env:PATH += ":/foo/bar"
or 
$env:PATH = $env:PATH + ':/foo/bar'
Would you expect the user to type something like the following:
'$env:PATH' = '$env:PATH' + ':/foo/bar'